core/node/node_sync/src/tree_data_fetcher/provider/mod.rs

```diff
@@ -0,0 +1,321 @@
+        if earliest_timestamp > l1_batch_seal_timestamp {
+            return Ok(earliest_number); // No better estimate at this point
+        }
+
+        // At this point, we have `earliest_timestamp <= l1_batch_seal_timestamp <= latest_timestamp`.
+        // Binary-search the range until we're sort of accurate.
+        let mut left = earliest_number;
+        let mut right = latest_number;
+        while left + Self::L1_BLOCK_ACCURACY < right {
+            let middle = (left + right) / 2;
+            let (_, middle_timestamp) =
+                Self::get_block(eth_client, web3::BlockNumber::Number(middle)).await?;
+            if middle_timestamp <= l1_batch_seal_timestamp {
+                left = middle;
+            } else {
+                right = middle;
+            }
+        }
+        Ok(left)
+    }
+
+    /// Gets a block that should be present on L1.
+    async fn get_block(
+        eth_client: &DynClient<L1>,
+        number: web3::BlockNumber,
+    ) -> EnrichedClientResult<(U64, U256)> {
+        let block = eth_client.block(number.into()).await?.ok_or_else(|| {
+            let err = "block is missing on L1 RPC provider";
+            EnrichedClientError::new(ClientError::Custom(err.into()), "get_block")
+                .with_arg("number", &number)
+        })?;
+        let number = block.number.ok_or_else(|| {
+            let err = "block is missing a number";
+            EnrichedClientError::new(ClientError::Custom(err.into()), "get_block")
+                .with_arg("number", &number)
+        })?;
+        Ok((number, block.timestamp))
+    }
+
+    pub fn with_fallback(self, fallback: Box<dyn TreeDataProvider>) -> CombinedDataProvider {
+        CombinedDataProvider {
+            l1: Some(self),
+            fallback,
+        }
+    }
+}
+
+#[async_trait]
+impl TreeDataProvider for L1DataProvider {
+    async fn batch_details(
+        &mut self,
+        number: L1BatchNumber,
+    ) -> TreeDataFetcherResult<Result<H256, MissingData>> {
+        let l1_batch_seal_timestamp = self.l1_batch_seal_timestamp(number).await?;
+        let from_block = self.past_l1_batch.and_then(|info| {
+            assert!(
+                info.number < number,
+                "`batch_details()` must be called with monotonically increasing numbers"
+            );
+            let threshold_timestamp = info.l1_commit_block_timestamp + Self::L1_BLOCK_RANGE.as_u64() / 2;
+            if U256::from(l1_batch_seal_timestamp) > threshold_timestamp {
+                tracing::debug!(
+                    number = number.0,
+                    "L1 batch #{number} seal timestamp ({l1_batch_seal_timestamp}) is too far ahead \
+                     of the previous processed L1 batch ({info:?}); not using L1 batch info"
+                );
+                None
+            } else {
+                // This is an exact lower boundary: L1 batches are committed in order
```

Comment:
Just to double-check: Is this always true? Seems so after reading L1 contracts, but I may be missing something.

Reply:
Correct